### ISSUE TITLE
Return default value when mock disabled or ignored

### ISSUE THREAD

**Issue by grantramsay** [cpputest/cpputest#863]
Hi, I was following the [example on website](http://cpputest.github.io/mocking_manual.html#other_mock_support) about `returnIntValueOrDefault` but I can't get it to work.
e.g.

    TEST(none, returns_default_when_mock_disabled)
    {
        mock().disable();
        LONGS_EQUAL(5, mock().actualCall("function").returnIntValueOrDefault(5));
    }

> error: Failure in TEST(none, returns_default_when_mock_disabled)
expected <5 0x5>
but was  <0 0x0>

I get the same thing with `mock().ignoreOtherCalls();`.
Currently using version 32dd4276545dd947e166b6bc9ae646f1e2895973

Am I using this correctly?

**Comment by arstrube** [comment]
I think this is expected behavior. We should probably add test cases for this. The following should work when mocks are disabled:
```c++
TEST(none, returns_default_when_mock_disabled)
{
    mock().disable();
    LONGS_EQUAL(5, mock().returnIntValueOrDefault(5));
}
```

**Comment by arstrube** [comment]
The reason is that `mock().disable()` will cause instances of `MockCheckedActualCall` to be replaced by instances of `MockIgnoredActualCall`, which only has dummy methods.

I think there should probably be at least one test case somewhere to demonstrate this.

**Comment by arstrube** [comment]
TEST(none, mock_returns_default_when_mock_disabled_but_actualCall_doesnt)
{
    mock().disable();
    LONGS_EQUAL(5, mock().returnIntValueOrDefault(5));
    LONGS_EQUAL(0, mock().actualCall("function").returnIntValueOrDefault(5));
}

**Comment by grantramsay** [comment]
So is that expected behavior?

The [example on website](http://cpputest.github.io/mocking_manual.html#other_mock_support) says:
> int function () {
    return mock().actualCall("function").returnIntValueOrDefault(5);
}
If the function is ignored or the framework is disabled, a 5 will be returned, otherwise the expected return value will be returned (according to the expectations set on the test case).

Return value logic like this would seem nice to me:

    if mock_enabled && expected_value_set
        return expected_value
    else if default_value_set
        return default_value
    else
        return 0 // or NULL etc.

**Comment by arstrube** [comment]
Hmm. Clearly, the documentation and the real behavior are different. Let's hear out @basvodde which one should be fixed. Fixing the behavior won't be as easy as the above code suggests.

**Comment by basvodde** [comment]
I guess that the IgnoredActualCall can *at least* return the default :) Perhaps simply not overriding the methods might do?

**Comment by arstrube** [comment]
Wishful thinking :-)

1. It needs to be a descendent of `MockCheckedExpectedCall` to do that, which is ugly to start with.

2. Making it so results in a test failure, provoked by a memory "leak", because it's statically allocated as a singleton, and presumably the memory allocated in its constructor (which, incidentally, it doesn't need) never gets deallocated before the memory leak check is done.

**Comment by basvodde** [comment]
AFAIK, the default would just return its parameter, so that wouldn't he hard....

**Comment by arstrube** [comment]
@basvodde there _may_ be a way of doing it, and it _may_ be simple, but deriving from MockCheckedActualCall is pretty ugly and it doesn't work right

**Comment by arstrube** [comment]
Actually, since an ignored call _always_ has to return the default (if one was supplied), it can be fixed directly in the header.

**Comment by arstrube** [comment]
There is a test failure as a result, but that failure is due to the wrong behavior being expected.

**Comment by basvodde** [comment]
This is the current  method:

```
    virtual double returnDoubleValueOrDefault(double) _override { return returnDoubleValue(); }
```

If we return the parameter instead, then it will always return the default. No need for any state for that. It isn't great, but it would be better than the current default behavior, I'd say.

**Comment by arstrube** [comment]
You are right. Why isn't it great?

**Comment by basvodde** [comment]
It doesn't work for non-default parameters :)

**Comment by arstrube** [comment]
But it doesn't have to. Those guys are undefined, and everybody ought to know that :)

**Comment by arstrube** [comment]
Fixed by PR #867

**Comment by arstrube** [comment]
@basvodde close this?
